site/content/blog/2022-03-22-schema-migrations.md

```diff
@@ -0,0 +1,130 @@
+
+```none
+0000000001_create_user.up.sql
+0000000001_create_user.down.sql
+```
+
+Apart from the essential arguments, morph can be initialized with a locking option which enables a lock mechanism (to mimic advisory locking) that’s been builtin with the tool. The reason we avoid using the advisory lock is that; we use [PgBouncer](https://github.com/pgbouncer/pgbouncer) on our cloud infrastructure and for scaling reasons we had to use transaction pooling mode instead of session pooling. The transaction pooling prevents using session-level advisory locks hence we had to implement our own locking feature to overcome this. In essence, this locking mechanism is being used for cases where multiple applications could execute the migrations at the same time. If the database driver implements the following interface, morph can be used safely in multi-application deployments:
+
+```Go
+type Locker interface {
+    Lock(context.Context) error
+    Unlock() error
+}
+```
+
+It essentially allows an instance to create a mutex value with expiration in the database and refreshes it until the instance finishes applying migrations. Other instances will keep waiting until the mutex is either removed from the database or get expired. The main difference here is that locking is not used for a single statement but it’s being used for the entire migration process. And this mechanism creates another level above sessions hence it wouldn’t fail if the advisory locks are bound to the sessions.
+
+The morph CLI provides the same functionality as the command line. Feel free to install and experiment as we think the command discovery should be sufficient to start using it!
+
+For those who have read this far, the following question may immediately come to mind: why you just didn't use an existing tool? Or at least, there should be some tools already developed for such problems (schema migrations). Yes, you are absolutely correct. Let us explain why we didn't use [golang-migrate](https://github.com/golang-migrate/migrate) which is one of the most popular projects and we already appreciate it. The main reason is a migration can fail at any time for various reasons, in a failure scenario `golang-migrate` leaves the migration state dirty and it requires some manual steps to fix the issue. Since we have idempotent migrations, it is OK to run a migration over and over again. So we don't need to use the dirty state. The application can even choose to roll back the migration in case of failure. The other motivation is to have a more granular schema version as we store every applied migration. Let's think of a scenario where you are planning to introduce 2 new migrations for the new release (let their version number be 18 and 19). At the last minute, you decided to pull out 18. migration. In this case, you need to rename 19. migration to 18 or rename the 18. migration to 20. but the way we store migrations do not require this. You can simply add 19. migration before the 18. one.
```

Comment:
What do you mean with "and we already appreciate it"?

Reply:
Minor: some odd use of "." after migration number e.g. "18. migration" "19. migration" do you mean "the 18th migration" and "the 19th migration"?

Reply:
> You can simply add 19. migration before the 18. one.

I don't understand, how will it know to execute this out of order? They are ordered by number right?

Reply:
I mean by appreciating it was; we actually used it, and still using on some other projects (eg. `mattermost-mattermod`). So we didn't actually completely moved away from using it.

For the ordering, yes the versions are there but you don't necessarily need to follow the order. I mean, it will be applied in order, but what if there is not a 18th migration existing? It will jump to 19th and you can apply 18th when it's ready, lets say on the next version.